ksqldb-engine/src/main/java/io/confluent/ksql/function/FunctionLoaderUtils.java

```diff
@@ -180,17 +180,27 @@ static ParamType getReturnType(
   static SchemaProvider handleUdfReturnSchema(
       final Class theClass,
       final ParamType javaReturnSchema,
+      final String annotationSchema,
+      final SqlTypeParser parser,
       final String schemaProviderFunctionName,
       final String functionName,
       final boolean isVariadic
   ) {
     final Function<List<SqlType>, SqlType> schemaProvider;
-    if (!schemaProviderFunctionName.equals("")) {
+    if (!Udf.NO_SCHEMA_PROVIDER.equals(schemaProviderFunctionName)) {
       schemaProvider = handleUdfSchemaProviderAnnotation(
           schemaProviderFunctionName, theClass, functionName);
-    } else if (javaReturnSchema instanceof DecimalType) {
-      throw new KsqlException(String.format("Cannot load UDF %s. BigDecimal return type "
-          + "is not supported without a schema provider method.", functionName));
+    } else if (!Udf.NO_SCHEMA.equals(annotationSchema)) {
+      schemaProvider = args -> parser.parse(annotationSchema).getSqlType();
+    } else if (!GenericsUtil.hasGenerics(javaReturnSchema)) {
+      final SqlType sqlType;
+      try {
+        sqlType = SchemaConverters.functionToSqlConverter().toSqlType(javaReturnSchema);
+      } catch (final Exception e) {
+        throw new KsqlException("Cannot load UDF " + functionName + ". "
+            + javaReturnSchema + " return type is not supported without a schema annotation.");
```

Comment:
Either a schema _or schemaProvider_ is required, right?

... and users may find 'schema annotation;' confusing, given the 'schema' is the name of a property of the annotation, not the name of the annotation.

Maybe:

```suggestion
        throw new KsqlException("Cannot load UDF " + functionName + ". "
            + javaReturnSchema + " return type is not supported without either a schema or schemaProvider set in the annotation.");
```

Though even better would be to pass down the type of the annotation so that the error could be explicit and say `set in the @Udf annotation`, or `in the @Udtf annotation` etc.

Reply:
I changed the error to clarify, but did not pass down the annotation. I think it's clear enough without it :)